staging/vhost-device-gpu/src/main.rs

```diff
@@ -34,28 +37,93 @@ pub mod gnu_main {
+        /// [gfxstream] Vulkan implementation
+        GfxstreamVulkan = GpuCapset::GFXSTREAM_VULKAN.bits(),
+        /// [virglrenderer] Vulkan implementation
+        Venus = GpuCapset::VENUS.bits(),
+        /// [virglrenderer] unsuported/experimental: DRM native context support
+        Drm = GpuCapset::DRM.bits(),
+        /// [gfxstream] unsuported/experimental
+        GfxstreamMagma = GpuCapset::GFXSTREAM_MAGMA.bits(),
+        /// [gfxstream] unsuported/experimental: OpenGL ES implementation
+        GfxstreamGles = GpuCapset::GFXSTREAM_GLES.bits(),
+        /// [gfxstream] unsuported/experimental
+        GfxstreamComposer = GpuCapset::GFXSTREAM_COMPOSER.bits(),
+    }
+
+    impl From<CapsetName> for GpuCapset {
+        fn from(capset_name: CapsetName) -> GpuCapset {
+            GpuCapset::from_bits(capset_name as u64)
+                .expect("Internal error: CapsetName enum is incorrectly defined")
+        }
+    }
+
+    pub fn capset_names_into_capset(
+        capset_names: impl IntoIterator<Item = CapsetName>,
+    ) -> GpuCapset {
+        capset_names
+            .into_iter()
+            .map(CapsetName::into)
+            .fold(GpuCapset::empty(), GpuCapset::union)
+    }
+
     #[derive(Parser, Debug)]
     #[clap(author, version, about, long_about = None)]
     pub struct GpuArgs {
         /// vhost-user Unix domain socket.
         #[clap(short, long, value_name = "SOCKET")]
-        pub socket_path: PathBuf,
-        #[clap(short, long, value_enum)]
-        pub gpu_mode: GpuMode,
-    }
+        socket_path: PathBuf,
 
-    impl From<GpuArgs> for GpuConfig {
-        fn from(args: GpuArgs) -> Self {
-            let socket_path = args.socket_path;
-            let gpu_mode: GpuMode = args.gpu_mode;
-
-            GpuConfig::new(socket_path, gpu_mode)
-        }
+        /// The mode specifies which backend implementation to use
+        #[clap(short, long, value_enum)]
+        gpu_mode: GpuMode,
+
+        /// Comma separated list of enabled capsets
+        #[clap(short, long, value_delimiter = ',')]
+        capset: Option<Vec<CapsetName>>,
+
+        /// Enable backend to use EGL
+        #[clap(
+            long,
+            action=ArgAction::Set,
+            default_value_t = GpuFlags::new_default().use_egl
+        )]
+        use_egl: bool,
+
+        /// Enable backend to use GLX
+        #[clap(
+            long,
+            action=ArgAction::Set,
+            default_value_t = GpuFlags::new_default().use_glx
+        )]
+        use_glx: bool,
+
+        /// Enable backend to use GLES
+        #[clap(
+            long,
+            action=ArgAction::Set,
+            default_value_t = GpuFlags::new_default().use_gles
+        )]
+        use_gles: bool,
+
+        /// Enable surfaceless backend option
+        #[clap(
+            long,
+            action = ArgAction::Set,
+            default_value_t = GpuFlags::new_default().use_surfaceless
+        )]
+        use_surfaceless: bool,
```

Comment:
how about maintaining the GpuArgs struct and instead do this to reduce the complexity:
```suggestion

    #[derive(Parser, Debug)]
    pub struct GpuArgs {
        #[clap(short, long, value_name = "SOCKET")]
        socket_path: PathBuf,

        #[clap(short, long, value_enum)]
        gpu_mode: GpuMode,

        #[clap(short, long, value_delimiter = ',')]
        capset: Option<Vec<CapsetName>>,

        #[clap(flatten)]
        flags: GpuFlagsArgs,
    }

    #[derive(Parser, Debug)]
    pub struct GpuFlagsArgs {
        #[clap(long, action = ArgAction::Set, default_value_t = GpuFlags::default().use_egl)]
        use_egl: bool,

        #[clap(long, action = ArgAction::Set, default_value_t = GpuFlags::default().use_glx)]
        use_glx: bool,

        #[clap(long, action = ArgAction::Set, default_value_t = GpuFlags::default().use_gles)]
        use_gles: bool,

        #[clap(long, action = ArgAction::Set, default_value_t = GpuFlags::default().use_surfaceless)]
        use_surfaceless: bool,
    }

    impl From<GpuFlagsArgs> for GpuFlags {
        fn from(args: GpuFlagsArgs) -> Self {
            GpuFlags {
                use_egl: args.use_egl,
                use_glx: args.use_glx,
                use_gles: args.use_gles,
                use_surfaceless: args.use_surfaceless,
            }
        }
    }